kotlin/samples/dungeon/common/src/main/java/com/squareup/sample/dungeon/PlayerWorkflow.kt

```diff
@@ -34,17 +34,27 @@ class PlayerWorkflow(
   private val cellsPerSecond: Float = 15f
 ) : StatefulWorkflow<ActorInput, Movement, Nothing, Rendering>() {
 
-  sealed class Event {
-    data class StartMoving(val direction: Direction) : Event()
-    data class StopMoving(val direction: Direction) : Event()
+  sealed class Action : WorkflowAction<Movement, Nothing> {
+    class StartMoving(val direction: Direction) : Action()
+
+    class StopMoving(val direction: Direction) : Action()
+
+    override fun Mutator<Movement>.apply(): Nothing? {
+      when (this@Action) {
```

Comment:
Ugh this is gross, didn't think about how this would interact with the receiver. Don't have any better ideas though

Reply:
I was just slavishly doing the Swift thing — this is their only option. In another spot I have each action implement it. Leagues better, will adjust this.

Reply:
It's a little verbose, but super easy to type thanks to the IDE, and easy to read.

```
  sealed class Action : WorkflowAction<Movement, Nothing> {
    class StartMoving(val direction: Direction) : Action() {
      override fun Mutator<Movement>.apply(): Nothing? {
        state += direction
        return null
      }
    }

    class StopMoving(val direction: Direction) : Action() {
      override fun Mutator<Movement>.apply(): Nothing? {
        state -= direction
        return null
      }
    }
  }
```

Reply:
Backed away from this and went back to `when (this@Action)`. It's really the smallest amount of boilerplate. 

Main wart is occasional cast like `val oldState = state`, but that pretty consistently seems like a clue that a workflow is too big and should be factored down into screen-sized bites.